parlai/agents/hugging_face/gpt2.py

```diff
@@ -0,0 +1,210 @@
+            # first step
+            if input.size(1) == 1 and int(input[0][0]) == self.start_idx:
+                # generating: ignore the start token
+                model_input = encoder_state
+            else:
+                # forced decoding: concatenate the context
+                # with the labels
+                model_input, _ = concat_without_padding(
+                    encoder_state,
+                    input,
+                    use_cuda=self.use_cuda,
+                    null_idx=self.null_idx,
+                )
+                attention_mask = model_input != self.null_idx
+        else:
+            # generation: get the last token input
+            model_input = input[:, -1].unsqueeze(1)
+
+        transformer_outputs = self.transformer(
+            model_input, past=incr_state, attention_mask=attention_mask
+        )
+        hidden_states = transformer_outputs[0]
+        new_incr_state = transformer_outputs[1]
+
+        return hidden_states, new_incr_state
+
+
+class HFGPT2Model(TorchGeneratorModel):
+    def __init__(self, opt, dict):
+        self.null_idx, self.start_idx, self.end_idx = self._get_special_tokens(
+            opt, dict
+        )
+        super().__init__(self.null_idx, self.start_idx, self.end_idx)
+
+        # init the model
+        self.encoder = DummyEncoder()
+        self.decoder = GPT2Decoder(opt, dict)
+
+        self.config = self.decoder.transformer.config
+        self.lm_head = torch.nn.Linear(
+            self.config.n_embd, self.config.vocab_size, bias=False
+        )
+        self.tie_weights(self.lm_head, self.decoder.transformer.wte)
+
+        self.text_lengths = None  # used to reverse concatenation of context and labels
+
+    def tie_weights(self, output_embeddings, input_embeddings):
+        output_embeddings.weight = input_embeddings.weight
+
+    def _get_special_tokens(self, opt, dict):
+        return dict.null_idx, dict.start_idx, dict.end_idx
+
+    def reorder_encoder_states(self, encoder_states, indices):
+        enc = torch.index_select(encoder_states, 0, indices)
+        return enc
+
+    def output(self, tensor):
+        """
+        Compute output logits.
+        """
+        # get only scores for labels
+        if self.text_lengths is not None:
+            total_length = max(self.text_lengths)
+            to_select = tensor.size(1) - total_length + 1
+            if to_select > 0:
+                # select only label scores
+                bsz = tensor.size(0)
+                new_tensors = []
+                for i in range(bsz):
+                    start = self.text_lengths[i] - 1
+                    end = start + to_select
+                    new_tensors.append(tensor[i, start:end, :].unsqueeze(0))
+                tensor = torch.cat(new_tensors, 0)
+
+        return self.lm_head(tensor)
+
+    def reorder_decoder_incremental_state(self, incremental_state, inds):
+        new_incr_state = []
+        for layer_past in incremental_state:
```

Comment:
I don't *know*, but this seems more expensive then it needs to be? why not a `torch.index_select` on each layer? 

The clone/detach is also particularly worrisome -- this would cause subtle issues if we did RL here and tried backpropping based on some RL signal.

Reply:
I just copied what they did, but can chnage